README.md
# Overview

**Reaver** has been designed to be a robust and practical attack against **Wi-Fi Protected Setup (WPS)** registrar PINs in order to **recover WPA/WPA2 passphrases**. It has been tested against a wide variety of access points and WPS implementations.

The **original** Reaver implements a **online brute force attack** against, as described in [http://sviehb.files.wordpress.com/2011/12/viehboeck_wps.pdf](http://sviehb.files.wordpress.com/2011/12/viehboeck_wps.pdf).
**reaver-wps-fork-t6x** is a **community forked version**, which has included **various bug fixes** and additional attack method (the **offline Pixie Dust** attack).

**Depending on the target's Access Point (AP)**, to recover the plain text WPA/WPA2 passphrase the **average** amount of time for the transitional **online brute force** method is **between 4-10 hours**. In practice, it will generally take half this time to guess the correct WPS pin and recover the passphrase.
When using the **offline attack**, **if** the AP is vulnerable, it may take only a matter of **seconds to minutes**.

* The original Reaver (v1.4) can be found here: [https://code.google.com/p/reaver-wps/](https://code.google.com/p/reaver-wps/).
* The discontinued community edition of Reaver (v1.5) that was used as the starting point: [https://code.google.com/p/reaver-wps/](https://code.google.com/p/reaver-wps/).
* reaver-wps-fork-t6x community edition of Reaver (which includes the Pixie Dust attack): [https://github.com/t6x/reaver-wps-fork-t6x](https://github.com/t6x/reaver-wps-fork-t6x).
* For more information about the Pixie Dust attack (including **which APs are vulnerable**) can be found here: 			[https://github.com/wiire/pixiewps](https://github.com/wiire/pixiewps), 
[https://forums.kali.org/showthread.php?24286-WPS-Pixie-Dust-Attack-(Offline-WPS-Attack)](https://forums.kali.org/showthread.php?24286-WPS-Pixie-Dust-Attack-(Offline-WPS-Attack)) & 									[https://docs.google.com/spreadsheets/d/1tSlbqVQ59kGn8hgmwcPTHUECQ3o9YhXR91A_p7Nnj5Y/edit?usp=sharing](https://docs.google.com/spreadsheets/d/1tSlbqVQ59kGn8hgmwcPTHUECQ3o9YhXR91A_p7Nnj5Y/edit?usp=sharing)

- - -

# Requirements

```
apt-get -y install build-essential libpcap-dev sqlite3 libsqlite3-dev aircrack-ng pixiewps
```
_The example uses [Kali Linux](https://www.kali.org/) as the Operating System (OS) as `pixiewps` is included._

You **must** already have Wiire's Pixiewps installed.
The latest version can be found here: [https://github.com/wiire/pixiewps](https://github.com/wiire/pixiewps).

- - -

# Setup

**Download**

`git clone https://github.com/t6x/reaver-wps-fork-t6x`

or

`wget https://github.com/t6x/reaver-wps-fork-t6x/archive/master.zip && unzip master.zip`

**Build**

```bash
cd reaver-wps-fork-t6x*/
cd src/
./configure
make
```

**Install**

`sudo make install`

- - -

# Reaver Usage

```
Reaver v1.5.2 WiFi Protected Setup Attack Tool
Copyright (c) 2011, Tactical Network Solutions, Craig Heffner <cheffner@tacnetsol.com>
mod by t6_x <t6_x@hotmail.com> & DataHead & Soxrok2212 & Wiire & kib0rg

Required Arguments:
		-i, --interface=<wlan>          Name of the monitor-mode interface to use
		-b, --bssid=<mac>               BSSID of the target AP

Optional Arguments:
		-m, --mac=<mac>                 MAC of the host system
		-e, --essid=<ssid>              ESSID of the target AP
		-c, --channel=<channel>         Set the 802.11 channel for the interface (implies -f)
		-o, --out-file=<file>           Send output to a log file [stdout]
		-s, --session=<file>            Restore a previous session file
		-C, --exec=<command>            Execute the supplied command upon successful pin recovery
		-D, --daemonize                 Daemonize reaver
		-a, --auto                      Auto detect the best advanced options for the target AP
		-f, --fixed                     Disable channel hopping
		-5, --5ghz                      Use 5GHz 802.11 channels
		-v, --verbose                   Display non-critical warnings (-vv for more)
		-q, --quiet                     Only display critical messages
		-K  --pixie-dust=<number>       [1] Run pixiewps with PKE, PKR, E-Hash1, E-Hash2, E-Nonce and Authkey (Ralink, Broadcom & Realtek)
		-Z, --no-auto-pass              Do NOT run reaver to auto retrieve WPA password if pixiewps attack is successful
		-h, --help                      Show help

Advanced Options:
		-p, --pin=<wps pin>             Use the specified 4 or 8 digit WPS pin
		-d, --delay=<seconds>           Set the delay between pin attempts [1]
		-l, --lock-delay=<seconds>      Set the time to wait if the AP locks WPS pin attempts [60]
		-g, --max-attempts=<num>        Quit after num pin attempts
		-x, --fail-wait=<seconds>       Set the time to sleep after 10 unexpected failures [0]
		-r, --recurring-delay=<x:y>     Sleep for y seconds every x pin attempts
		-t, --timeout=<seconds>         Set the receive timeout period [5]
		-T, --m57-timeout=<seconds>     Set the M5/M7 timeout period [0.20]
		-A, --no-associate              Do not associate with the AP (association must be done by another application)
		-N, --no-nacks                  Do not send NACK messages when out of order packets are received
		-S, --dh-small                  Use small DH keys to improve crack speed
		-L, --ignore-locks              Ignore locked state reported by the target AP
		-E, --eap-terminate             Terminate each WPS session with an EAP FAIL packet
		-n, --nack                      Target AP always sends a NACK [Auto]
		-w, --win7                      Mimic a Windows 7 registrar [False]
		-X, --exhaustive                Set exhaustive mode from the beginning of the session [False]
		-1, --p1-index                  Set initial array index for the first half of the pin [False]
		-2, --p2-index                  Set initial array index for the second half of the pin [False]
		-P, --pixiedust-loop            Set into PixieLoop mode (doesn't send M4, and loops through to M3) [False]
		-W, --generate-pin              Default Pin Generator by devttys0 team [1] Belkin [2] D-Link
		-H, --pixiedust-log             Enables logging of sequence completed PixieHashes

Example:
		reaver -i wlan0mon -b 00:AA:BB:11:22:33 -vvv -K 1
```

## -K // --pixie-dust

The `-K 1` option performances the offline attack, Pixie Dust _(`pixiewps`)_, by automatically passing the **PKE**, **PKR**, **E-Hash1**, **E-Hash2**, **E-Nonce** and **Authkey** variables. `pixiewps` will then try to attack **Ralink**, **Broadcom** and **Realtek** detected chipset.
**Special note**: If you are attacking a **Realtek AP**, **do NOT** use small DH Keys (`-S`) option.

## -H // --pixiedust-log

The `-H` option is a switch to enable logging of PixieHashes, **saved hashes** will be saved in the **executing directory**.
This option requires you to have at the least `-vvv` switch on and will work with `-K 1` & `-P` appropriately.

The files saved are named after the bssid (MAC) of the target, and with an extention of `.pixie`.
On the inside of these saved logs, you will find all the required **PixieDust hashes**, along with a **quick copy & paste ready full command** to use it on `pixiewps`. You also have the option to execute it. Just pop in the file into your favorite shell, and execute it _(`chmod +x <filename>` may be required)_.

## -P // --pixiedust-loop

When using the `-P` option, Reaver goes into a loop mode that breaks the WPS protocol by not using M4 message to **hopefully avoid lockouts**.
This is to **ONLY** be used for PixieHash collecting to use with `pixiewps`, **NOT** to brute forcing 'online' pins.

This option was made with intent of:

* Collecting repetitive hashes for further comparison and or analysis / **discovery of new vulnerable chipsets**, routers etc..
* **Time sensitive attacks** where the hash collecting continues repetitively until your time frame is met.
* For **scripting purposes** of whom want to use a possible lockout preventable way of PixieHash gathering for your use case.

- - -

# Wash Usage

```
Wash v1.5.2 WiFi Protected Setup Scan Tool
Copyright (c) 2011, Tactical Network Solutions, Craig Heffner <cheffner@tacnetsol.com>
mod by t6_x <t6_x@hotmail.com> & DataHead & Soxrok2212 & Wiire & kib0rg

Required Arguments:
		-i, --interface=<iface>              Interface to capture packets on
		-f, --file [FILE1 FILE2 FILE3 ...]   Read packets from capture files

Optional Arguments:
		-c, --channel=<num>                  Channel to listen on [auto]
		-o, --out-file=<file>                Write data to file
		-n, --probes=<num>                   Maximum number of probes to send to each AP in scan mode [15]
		-D, --daemonize                      Daemonize wash
		-C, --ignore-fcs                     Ignore frame checksum errors
		-5, --5ghz                           Use 5GHz 802.11 channels
		-s, --scan                           Use scan mode
		-u, --survey                         Use survey mode [default]
		-P, --file-output-piped              Allows Wash output to be piped. Example. wash x|y|z...
		-g, --get-chipset                    Pipes output and runs reaver alongside to get chipset
		-h, --help                           Show help

Example:
		wash -i wlan0mon
```

## -g // --get-chipset

The option `-g` of Wash, automatically runs Reaver to receive the chipset data.
**If** the AP does not respond to them quickly, this option will be **slow to display the data**, because Reaver will stay running until getting the data or until you reach your timeout limit (30 seconds).

- - -

# Acknowledgements

## Contribution

Modifications made by:
`t6_x`, `DataHead`, `Soxrok2212`, `Wiire`, `kib0rg`

Some ideas made by:
`nuroo`, `kcdtv`

Bug fix made by:
`alxchk`, `flatr0ze`, `USUARIONUEVO`, `ldm314`

## Special Thanks

* `Soxrok2212` for all work done to help in the development of tools
* `Wiire` for developing Pixiewps
* `Craig Heffner` for creating Reaver and for the creation of default pin generators (D-Link, Belkin) - http://www.devttys0.com/


CHANGELOG.md
## Change Log

## (31/05/2015)
- [a41ae4f](https://github.com/t6x/reaver-wps-fork-t6x/commit/a41ae4fd3ee4f9091d74a2ea8f2a9f4227b66c9c) Bug Fix, Pixewps mutual exclusivity with -r and -S options (@ldm314)

## (15/05/2015)
- [63c5c02](https://github.com/t6x/reaver-wps-fork-t6x/commit/63c5c02bfc44e732399118d844906b0708e8e28c) Bug Fix, reaver -g option (@t6x)
- [63c5c02](https://github.com/t6x/reaver-wps-fork-t6x/commit/63c5c02bfc44e732399118d844906b0708e8e28c) Bug Fix, reaver -p option (@t6x)
- [63c5c02](https://github.com/t6x/reaver-wps-fork-t6x/commit/63c5c02bfc44e732399118d844906b0708e8e28c) Bug Fix, fix M7 Message (@t6x)
- [63c5c02](https://github.com/t6x/reaver-wps-fork-t6x/commit/63c5c02bfc44e732399118d844906b0708e8e28c) Bug Fix, fix time out in M3 or M5 message (@t6x)
- [63c5c02](https://github.com/t6x/reaver-wps-fork-t6x/commit/63c5c02bfc44e732399118d844906b0708e8e28c) Wash Ignore FCS by default (-C option)(@t6x)
- [63c5c02](https://github.com/t6x/reaver-wps-fork-t6x/commit/63c5c02bfc44e732399118d844906b0708e8e28c) Credits (@t6x)
- [63c5c02](https://github.com/t6x/reaver-wps-fork-t6x/commit/63c5c02bfc44e732399118d844906b0708e8e28c) Change Log (@t6x)

## (11/05/2015)
- [ead842c](https://github.com/t6x/reaver-wps-fork-t6x/commit/ead842cf17a160e6fcc0f396066c4da758912593) Bug Fix (@t6x)

## (04/05/2015)
- [68515bd](https://github.com/t6x/reaver-wps-fork-t6x/commit/68515bd90d9c952992bfe0d152c069a8e25521bf) Improved code (@t6x)
- [a28b0af](https://github.com/t6x/reaver-wps-fork-t6x/commit/a28b0af2a909394156fa699a1fa3750d10ff0613) Improved code, updating to use -f option of pixiewps (@t6x)
- [714fdcf](https://github.com/t6x/reaver-wps-fork-t6x/commit/714fdcf005be92dc58db31ca1a51f5c27ea19d4f) Updated build reaver instructions (@soxrok2212)
- [c69503a](https://github.com/t6x/reaver-wps-fork-t6x/commit/c69503a4d485b06ca601329332cde67c64a8aa55) Update wpsmon.c (@t6x)
- [ab303ae](https://github.com/t6x/reaver-wps-fork-t6x/commit/ab303aeb4aaef04709c741e880525c635b7dc8ce) Update wpscrack.c (@t6x)
- [679bf56](https://github.com/t6x/reaver-wps-fork-t6x/commit/679bf56e1d6a3436adf15a78f3dbff8c032ee62b) Update README.md (@t6x)

## (02/05/2015)
- [a29e53f](https://github.com/t6x/reaver-wps-fork-t6x/commit/a29e53fdc3f7520aa7228ad3438d5054676af55d) Fixed bad pin parsing (@wiire)

## (28/04/2015)
- [88b0a90](https://github.com/t6x/reaver-wps-fork-t6x/commit/88b0a90d0f871be118ac8f96fa676b5be7c56ea3) Update wps_registrar.c (@d8tahead)
- [61a96dc](https://github.com/t6x/reaver-wps-fork-t6x/commit/61a96dc8b5b41e386591e36a758f6b9711218557) Add R-Nonce and Model Name output (@t6x)
- [485a6b8](https://github.com/t6x/reaver-wps-fork-t6x/commit/485a6b8b7bbac8433ddf4b04f1cd1a885e3d6261) Add Model Name in output (@t6x)

## (27/04/2015)
- [cccae8d](https://github.com/t6x/reaver-wps-fork-t6x/commit/cccae8d0a0b0f980c29861bec02152ac07d505e5) Update wps_registrar.c (@d8tahead)
- [34583f6](https://github.com/t6x/reaver-wps-fork-t6x/commit/34583f62416e29f4226c2769e524607feb5f6b86) Update wps_registrar.c (@d8tahead)

## (26/04/2015)
- [086e4d9](https://github.com/t6x/reaver-wps-fork-t6x/commit/086e4d9d5aaa166940c9e507edaf4e359a2ed59a) Added [P] R-Nonce for future pixiewps use :) (@d8tahead)

## (21/04/2015)
- [94cd799](https://github.com/t6x/reaver-wps-fork-t6x/commit/94cd79967d09eedd43f151b3992e335ecff7479d) Update globule.c (@d8tahead)
- [2166159](https://github.com/t6x/reaver-wps-fork-t6x/commit/2166159f8534c46d8d9250558fb9bd36c0aa9aae) Update globule.h (@d8tahead)
- [866919b](https://github.com/t6x/reaver-wps-fork-t6x/commit/866919b9cf08e19e08152b422ddd0a5d4b42f494) Update exchange.c (@d8tahead)
- [f767655](https://github.com/t6x/reaver-wps-fork-t6x/commit/f767655b0254758048aed44bfb77981d66028a20) Serial number printf fix (@soxrok2212)

## (20/04/2015)
- [ae579a1](https://github.com/t6x/reaver-wps-fork-t6x/commit/ae579a1eef4f5e00958c44ae9bc1d1af6e0ab1f1) Update README.md (@d8tahead)

## (19/04/2015)
- [2e41e9f](https://github.com/t6x/reaver-wps-fork-t6x/commit/2e41e9fc1fc48b2beab1a2c99a56b705712d7a7d) Special realtek note (@soxrok2212)
- [2e9c7af](https://github.com/t6x/reaver-wps-fork-t6x/commit/2e9c7af5b5c100b6dedec8651afa867cbe1541b2) Update README.md (@soxrok2212)
- [9310bab](https://github.com/t6x/reaver-wps-fork-t6x/commit/9310babf80988d9a955ca55aae7072fdb1fea39f) Small fix (@soxrok2212)
- [9358bf3](https://github.com/t6x/reaver-wps-fork-t6x/commit/9358bf38af5f0c254261745f05583c7242e5e0a7) A few more small fixes (@soxrok2212)
- [a2bdeda](https://github.com/t6x/reaver-wps-fork-t6x/commit/a2bdeda3987e0328c1782117a7eb7ec79b033ccf) Grammer and Craig Heffner credit (@soxrok2212)
- [b3be5fb](https://github.com/t6x/reaver-wps-fork-t6x/commit/b3be5fb95c86e577ac26bf06cb46f28a2d2052e3) Update wpsmon.c (@d8tahead)

## (18/04/2015)
- [e20ba7d](https://github.com/t6x/reaver-wps-fork-t6x/commit/e20ba7d7920b21e9dfb0d4f7a1ef564d86037087) Changes in pixiewps use, idea soxrok2212 (@t6x)
- [c6d2474](https://github.com/t6x/reaver-wps-fork-t6x/commit/c6d2474378d19c88cd1470e753a2c71aeeed58fc) Ideia by Soxrok2212, serial print in wash with option -g (@t6x)

## (17/04/2015)
- [7298416](https://github.com/t6x/reaver-wps-fork-t6x/commit/7298416d1102a2d546eee0fa72a6445b565a59de) Merge with DataHead branch (@t6x)

## (16/04/2015)
- [1a0994b](https://github.com/t6x/reaver-wps-fork-t6x/commit/1a0994b6b3b4b8424d8f31e8653ce010bf0d14e7) Credits devttys0 (@t6x)
- [37032a5](https://github.com/t6x/reaver-wps-fork-t6x/commit/37032a5b2d351b80ad472df5947987f156d1b7fd) Default pin generator by devttys0 (@t6x)

## (15/04/2015)
- [8d0ef55](https://github.com/t6x/reaver-wps-fork-t6x/commit/8d0ef55d244fb75138cce6e94921cf230f110b9b) Credits (@t6x)
- [d65da86](https://github.com/t6x/reaver-wps-fork-t6x/commit/d65da863530822d83627457a44a9e94ba1617ab6) Reaver serial number output verbose (@t6x)
- [433636e](https://github.com/t6x/reaver-wps-fork-t6x/commit/433636ef5e87a69d84531c15d312beeade597863) Update README.md (@t6x)

src/defs.h
/*
 * Reaver - Common definitions
 * Copyright (c) 2011, Tactical Network Solutions, Craig Heffner <cheffner@tacnetsol.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *
 *  In addition, as a special exception, the copyright holders give
 *  permission to link the code of portions of this program with the
 *  OpenSSL library under certain conditions as described in each
 *  individual source file, and distribute linked combinations
 *  including the two.
 *  You must obey the GNU General Public License in all respects
 *  for all of the code used other than OpenSSL. *  If you modify
 *  file(s) with this exception, you may extend this exception to your
 *  version of the file(s), but you are not obligated to do so. *  If you
 *  do not wish to do so, delete this exception statement from your
 *  version. *  If you delete this exception statement from all source
 *  files in the program, then also delete it here.
 */

#ifndef DEFS_H
#define DEFS_H

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <pcap.h>
#include <asm/byteorder.h>

#include "wps.h"

#define NULL_MAC		"\x00\x00\x00\x00\x00\x00"
#define DEFAULT_MAX_NUM_PROBES	15
#define MAX_ASSOC_FAILURES	10

#define TIMESTAMP_LEN		8
#define MAC_ADDR_LEN    	6
#define SSID_TAG_NUMBER		0
#define RATES_TAG_NUMBER	1
#define CHANNEL_TAG_NUMBER	3
#define WPS_TAG_NUMBER		0xDD
#define VENDOR_SPECIFIC_TAG	0xDD
#define RSN_TAG_NUMBER		0x30

#define CAPABILITY_WEP		0x10

#define WPA_IE_ID               "\x00\x50\xF2\x01\x01\x00"
#define WPA_IE_ID_LEN           6

#define MANAGEMENT_FRAME	0x00
#define SUBTYPE_BEACON		0x08

#define DOT1X_AUTHENTICATION	0x888E
#define DOT1X_EAP_PACKET	0x00

#define SIMPLE_CONFIG		0x00000001

#define P1_SIZE			10000
#define P2_SIZE			1000

#define EAPOL_START_MAX_TRIES	25
#define WARN_FAILURE_COUNT	10

#define EAPOL_START		1
#define EAP_IDENTITY 		0x01
#define EAP_EXPANDED            0xFE

#define M57_DEFAULT_TIMEOUT     200000          /* uSeconds */
#define M57_MAX_TIMEOUT         1000000         /* uSeconds */
#define DEFAULT_DELAY           1               /* Seconds */
#define DEFAULT_TIMEOUT         5               /* Seconds */
#define DEFAULT_LOCK_DELAY      60              /* Seconds */
#define SEC_TO_US               1000000         /* uSeconds in a Second */

#define TSFT_SIZE 		8
#define FLAGS_SIZE 		1
#define RATE_SIZE 		1
#define CHANNEL_SIZE 		4
#define FHSS_SIZE 		2

#define WPS_DEVICE_NAME		"Glau"
#define WPS_MANUFACTURER	"Microsoft"
#define WPS_MODEL_NAME		"Windows"
#define WPS_MODEL_NUMBER	"6.1.7601"
#define WPS_DEVICE_TYPE		"\x00\x01\x00\x50\xF2\x04\x00\x01"
#define WPS_OS_VERSION		"\x01\x00\x06\x00"
#define WPS_RF_BANDS		0x01

enum encryption_type
{
    NONE,
    WEP,
    WPA
};

enum key_state
{
    KEY1_WIP = 0,
    KEY2_WIP = 1,
    KEY_DONE = 2
};

enum debug_level
{
    CRITICAL = 0,
    INFO = 1,
    WARNING = 2,
    VERBOSE = 3
};

enum eap_codes
{
    EAP_REQUEST = 1,
    EAP_RESPONSE = 2,
    EAP_SUCCESS = 3,
    EAP_FAILURE = 4
};

enum wps_result
{
    KEY_ACCEPTED = 0,
    KEY_REJECTED = 1,
    RX_TIMEOUT = 2,
    EAP_FAIL = 3,
    UNKNOWN_ERROR = 4
};

enum nack_code
{
    NO_NACK = -1,
    NO_ERROR = 0,
    OOB_RRAD_ERROR = 1,
    CRC_FAILURE = 2,
    CHANNEL_24_NS = 3,
    CHANNEL_50_NS = 4,
    WEAK_SIGNAL = 5,
    NET_AUTH_FAILURE = 6,
    NET_ASSOCIATION_FAILURE = 7,
    NO_DHCP_RESPONSE = 8,
    FAILED_DHCP_CONFIG = 9,
    IP_ADDR_CONFLICT = 10,
    REGISTRAR_CONNECT_FAILURE = 11,
    MULTIPLE_PBC = 12,
    ROGUE_ACTIVITY = 13,
    DEVICE_BUSY = 14,
    SETUP_LOCKED = 15,
    MESSAGE_TIMEOUT = 16,
    REGISTRATION_TIMEOUT = 17,
    AUTH_FAILURE = 18
};

enum wps_type
{
    TERMINATE = -1,
    UNKNOWN = 0,
    IDENTITY_REQUEST = 1,
    IDENTITY_RESPONSE = 2,
    M1 = 0x04,
    M2 = 0x05,
    M3 = 0x07,
    M4 = 0x08,
    M5 = 0x09,
    M6 = 0x0A,
    M7 = 0x0B,
    M8 = 0x0C,
    DONE = 0x0F,
    NACK = 0x0E
};

enum rt_header_flags
{
    TSFT_FLAG = 0x01,
    FLAGS_FLAG = 0x02,
    RATE_FLAG = 0x04,
    CHANNEL_FLAG = 0x08,
    FHSS_FLAG = 0x10,
    SSI_FLAG = 0x20,
};

enum wfa_elements
{
    AP_CHANNEL = 0x1001,
    ASSOCIATION_STATE = 0x1002,
    AUTHENTICATION_TYPE = 0x1003,
    AUTHENTICATION_TYPE_FLAGS = 0x1004,
    AUTHENTICATOR = 0x1005,
    CONFIG_METHODS = 0x1008,
    CONFIGURATION_ERROR = 0x1009,
    CONFIRMATION_URL_4 = 0x100A,
    CONFIRMATION_URL_6 = 0x100B,
    CONNECTION_TYPE = 0x100C,
    CONNECTION_TYPE_FLAGS = 0x100D,
    CREDENTIAL = 0x100E,
    DEVICE_NAME = 0x1011,
    DEVICE_PASSWORD_ID = 0x1012,
    ENROLLEE_HASH_1 = 0x1014,
    ENROLLEE_HASH_2 = 0x1015,
    ENROLLEE_SNONCE_1 = 0x1016,
    ENROLLEE_SNONCE_2 = 0x1017,
    ENCRYPTED_SETTINGS = 0x1018,
    ENCRYPTION_TYPE = 0x100F,
    ENCRYPTION_TYPE_FLAGS = 0x1010,
    ENROLLEE_NONCE = 0x101A,
    FEATURE_ID = 0x101B,
    IDENTITY = 0x101C,
    IDENTITY_PROOF = 0x101D,
    KEY_WRAP_AUTHENTICATOR = 0x101E,
    KEY_IDENTIFIER = 0x101F,
    MAC_ADDRESS = 0x1020,
    MANUFACTURER = 0x1021,
    MESSAGE_TYPE = 0x1022,
    MODEL_NAME = 0x1023,
    MODEL_NUMBER = 0x1024,
    NETWORK_INDEX = 0x1026,
    NETWORK_KEY = 0x1027,
    NETWORK_KEY_INDEX = 0x1028,
    NEW_DEVICE_NAME = 0x1029,
    NEW_PASSWORD = 0x102A,
    OOB_DEVICE_PASSWORD = 0x102C,
    OS_VERSION = 0x102D,
    POWER_LEVEL = 0x102F,
    PSK_CURRENT = 0x1030,
    PSK_MAX = 0x1031,
    PUBLIC_KEY = 0x1032,
    RADIO_ENABLED = 0x1033,
    REBOOT = 0x1034,
    REGISTRAR_CURRENT = 0x1035,
    REGISTRAR_ESTABLISHED = 0x1036,
    REGISTRAR_LIST = 0x1037,
    REGISTRAR_MAX = 0x1038,
    REGISTRAR_NONCE = 0x1039,
    REQUEST_TYPE = 0x103A,
    RESPONSE_TYPE = 0x103B,
    RF_BANDS = 0x103C,
    REGISTRAR_HASH_1 = 0x103D,
    REGISTRAR_HASH_2 = 0x103E,
    REGISTRAR_SNONCE_1 = 0x103F,
    REGISTRAR_SNONCE_2 = 0x1040,
    SELECTED_REGISTRAR = 0x1041,
    SERIAL_NUMBER = 0x1042,
    WPS_STATE = 0x1044,
    SSID = 0x1045,
    TOTAL_NETWORKS = 0x1046,
    ENROLLEE_UUID = 0x1047,
    REGISTRAR_UUID = 0x1048,
    VENDOR_EXTENSION = 0x1049,
    VERSION = 0x104A,
    X509_CERT_REQUEST = 0x104B,
    X509_CERT = 0x104C,
    WPS_EAP_IDENTITY = 0x104D,
    MESSAGE_COUNTER = 0x104E,
    PUBLIC_KEY_HASH = 0x104F,
    REKEY_KEY = 0x1050,
    KEY_LIFETIME = 0x1051,
    PERMITTED_CONFIG_METHODS = 0x1052,
    SELECTED_REGISTRAR_CONFIG_METHODS = 0x1053,
    PRIMARY_DEVICE_TYPE = 0x1054,
    SECONDARY_DEVICE_TYPE_LIST = 0x1055,
    PORTABLE_DEVICE = 0x1056,
    AP_SETUP_LOCKED = 0x1057,
    APPLICATION_EXTENSION = 0x1058,
    EAP_TYPE = 0x1059,
    INITIALIZATION_VECTOR = 0x1060,
    KEY_PROVIDED_AUTOMATICALLY = 0x1061,
    ENABLED_8021X = 0x1062,
    APP_SESSION_KEY = 0x1063,
    WEP_TRANSMIT_KEY = 0x10064
};

#define IEEE80211_FCTL_VERS		0x0003
#define IEEE80211_FCTL_FTYPE		0x000c
#define IEEE80211_FCTL_STYPE		0x00f0
#define IEEE80211_FCTL_TODS		0x0100
#define IEEE80211_FCTL_FROMDS		0x0200
#define IEEE80211_FCTL_MOREFRAGS	0x0400
#define IEEE80211_FCTL_RETRY		0x0800
#define IEEE80211_FCTL_PM		0x1000
#define IEEE80211_FCTL_MOREDATA		0x2000
#define IEEE80211_FCTL_PROTECTED	0x4000
#define IEEE80211_FCTL_ORDER		0x8000

#define IEEE80211_SCTL_FRAG		0x000F
#define IEEE80211_SCTL_SEQ		0xFFF0

#define IEEE80211_FTYPE_MGMT		0x0000
#define IEEE80211_FTYPE_CTL		0x0004
#define IEEE80211_FTYPE_DATA		0x0008

/* management */
#define IEEE80211_STYPE_ASSOC_REQ	0x0000
#define IEEE80211_STYPE_ASSOC_RESP	0x0010
#define IEEE80211_STYPE_REASSOC_REQ	0x0020
#define IEEE80211_STYPE_REASSOC_RESP	0x0030
#define IEEE80211_STYPE_PROBE_REQ	0x0040
#define IEEE80211_STYPE_PROBE_RESP	0x0050
#define IEEE80211_STYPE_BEACON		0x0080
#define IEEE80211_STYPE_ATIM		0x0090
#define IEEE80211_STYPE_DISASSOC	0x00A0
#define IEEE80211_STYPE_AUTH		0x00B0
#define IEEE80211_STYPE_DEAUTH		0x00C0
#define IEEE80211_STYPE_ACTION		0x00D0

/* control */
#define IEEE80211_STYPE_BACK_REQ	0x0080
#define IEEE80211_STYPE_BACK		0x0090
#define IEEE80211_STYPE_PSPOLL		0x00A0
#define IEEE80211_STYPE_RTS		0x00B0
#define IEEE80211_STYPE_CTS		0x00C0
#define IEEE80211_STYPE_ACK		0x00D0
#define IEEE80211_STYPE_CFEND		0x00E0
#define IEEE80211_STYPE_CFENDACK	0x00F0

/* data */
#define IEEE80211_STYPE_DATA			0x0000
#define IEEE80211_STYPE_DATA_CFACK		0x0010
#define IEEE80211_STYPE_DATA_CFPOLL		0x0020
#define IEEE80211_STYPE_DATA_CFACKPOLL		0x0030
#define IEEE80211_STYPE_NULLFUNC		0x0040
#define IEEE80211_STYPE_CFACK			0x0050
#define IEEE80211_STYPE_CFPOLL			0x0060
#define IEEE80211_STYPE_CFACKPOLL		0x0070
#define IEEE80211_STYPE_QOS_DATA		0x0080
#define IEEE80211_STYPE_QOS_DATA_CFACK		0x0090
#define IEEE80211_STYPE_QOS_DATA_CFPOLL		0x00A0
#define IEEE80211_STYPE_QOS_DATA_CFACKPOLL	0x00B0
#define IEEE80211_STYPE_QOS_NULLFUNC		0x00C0
#define IEEE80211_STYPE_QOS_CFACK		0x00D0
#define IEEE80211_STYPE_QOS_CFPOLL		0x00E0
#define IEEE80211_STYPE_QOS_CFACKPOLL		0x00F0

#pragma pack(1)
struct radio_tap_header
{
    uint8_t revision;	
    uint8_t pad;
	__le16 len;
	__le32 flags;
};

struct dot11_frame_header
{
	__le16 fc;
	__le16 duration;
    unsigned char addr1[MAC_ADDR_LEN];
    unsigned char addr2[MAC_ADDR_LEN];
    unsigned char addr3[MAC_ADDR_LEN];
	__le16 frag_seq;
};

struct authentication_management_frame
{
	__le16 algorithm;
	__le16 sequence;
	__le16 status;
};

struct association_request_management_frame
{
	__le16 capability;
	__le16 listen_interval;
};

struct association_response_management_frame
{
	__le16 capability;
	__le16 status;
	__le16 id;
};

struct beacon_management_frame
{
    unsigned char timestamp[TIMESTAMP_LEN];
	__le16 beacon_interval;
	__le16 capability;
};

struct llc_header
{
    uint8_t dsap;
    uint8_t ssap;
    uint8_t control_field;
    unsigned char org_code[3];
	__be16 type;
};

struct dot1X_header
{
    uint8_t version;
    uint8_t type;
    uint16_t len;
};

struct eap_header
{
    uint8_t code;
    uint8_t id;
    uint16_t len;
    uint8_t type;
};

struct wfa_expanded_header
{
    unsigned char id[3];
	__be32 type;
    uint8_t opcode;
    uint8_t flags;
};

struct wfa_element_header
{
    uint16_t type;
    uint16_t length;
};

struct tagged_parameter
{
    uint8_t number;
    uint8_t len;
};
#pragma pack()

#define MIN_BEACON_SIZE		(sizeof(struct radio_tap_header) + sizeof(struct dot11_frame_header) + sizeof(struct beacon_management_frame))

#endif
